### PR TITLE
RFC 0587 Encryption Envelope v2

### PR DESCRIPTION
**Pull request by troyronda**
Signed-off-by: Troy Ronda <troy@troyronda.com>

### CODE REVIEW

**Review by llorllale — APPROVED** [review]

**Review by kdenhartog — APPROVED** [review]
Using this for AIP 2.0 is going to make the transition much smoother

**Review by Baha-sk — APPROVED** [review]

**Review by swcurran — CHANGES_REQUESTED** [review]
Could you add how we would move to this in existing frameworks?   Will we need a community update approach?  I imagine that same question goes for RFC 0334.  This one talks about what we should do, but not about how we get there.

Dummy question: It also looks like we are saying we all MUST include multiple crypto suites vs. negotiating what we support.  Is that right?

Thanks

**Comment by llorllale** [comment]
> negotiating what we support

I proposed #576 a while back. I'm not fond of its details, but the key point is in communicating the supported suites in the OOB invitation when bootstrapping a connection. This can be complemented with the Features Discovery protocol as well (as per the discussion over there).

**Comment by llorllale** [comment]
> Will we need a community update approach?

That depends on the community.

SecureKey is comfortable with jumping into this without intermediate steps.

**Comment by swcurran** [comment]
To clarify -- will we need a community update so that other frameworks that are deployed in the wild can move to this?  We need to keep interop going amongst the components for which interop is already working.

@andrewwhitehead - can you please take a look and assess how we would move to this?

**Comment by troyronda** [comment]
@swcurran Here's an example community update:

-  Implement new envelope format and algorithms.
- Add property to OOB to indicate envelope "version" (easier to define) or supported individual "suites" (harder to define).
- Frameworks that support the newer version use the newer version when the OOB property indicates the other side does as well.
- Transition out of supporting old format. (hopefully not too long of a timeline).

**Comment by troyronda** [comment]
@swcurran I have created a sibling PR: https://github.com/hyperledger/aries-rfcs/pull/595 that adds an `accept` property to the OOB message. Using the `accept` property in the OOB message, the acceptable envelope type(s) can be determined by the receiver of an OOB message.

**Comment by troyronda** [comment]
Updated with:

- renamed to encryption-envelope-v2.
- mime type clarification.
- commentary about the DIDComm v2 transition.

File: features/0587-encryption-envelope-v2/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+## Motivation
+
+This RFC defines ciphersuites for envelopes such that we can achieve better compatability with DIDComm Messaging being specified at DIF.
+The ciphersuites defined in this RFC are a subset of the definitions in [Aries RFC 0334-jwe-envelope](https://github.com/hyperledger/aries-rfcs/tree/master/features/0334-jwe-envelope).
+
+## Encryption Algorithms
+
+[ECDH-1PU draft 03](https://tools.ietf.org/html/draft-madden-jose-ecdh-1pu-03) defines the JWE structure. The following sections summarize the supported algorithms.
+
+### Curves
```
Comment (troyronda) [inline features/0587-encryption-envelope-v2/README.md:24]: DIDComm v2 requires at least X25519 and P-256 and also allows for P-384 and P-521.

Note - It’s worthwhile to also support P-384 and would be preferable for some implementers as compared to P-256.

**Comment by troyronda** [comment]
Added clarification on the suites required by this RFC and context on the DIDComm Messaging requirements.

File: features/0587-encryption-envelope-v2/README.md (hunk 1 above)
Comment (troyronda) [reply features/0587-encryption-envelope-v2/README.md:24]: Added clarification text to RFC.

File: features/0587-encryption-envelope-v2/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+## Anoncrypt equivalent
+
+For situations where the envelopes should be anonymous, we use a newly minted DID rather than a different Anoncrypt mechanism.
+The newly minted DID is used once and then discarded.
+This approach matches the DIDComm Messaging mechanism.
+
+## Mime Type
+
+The mime type associated to this envelope is `application/didcomm-encrypted+json`.
```
Comment (troyronda) [inline features/0587-encryption-envelope-v2/README.md:56]: The DIDComm Messaging spec currently states: `application/didcomm-enc-env`
https://identity.foundation/didcomm-messaging/spec/#https

RFC 0044 currently states (for v2): `application/didcomm-encrypted+json`

File: features/0587-encryption-envelope-v2/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+[RFC 0044](../0044-didcomm-file-and-mime-types/README.md) provides a general discussion of mime types.
+
+## DIDComm v2 Transition
+
+As this RFC specifies the same envelope format as will be used in DIDComm v2, an implementor should detect if the payload contains DIDComm v1 content or the JWM from DIDComm v2.
```
Comment (llorllale) [inline features/0587-encryption-envelope-v2/README.md:61]: https://github.com/decentralized-identity/didcomm-messaging/issues/165

File: features/0587-encryption-envelope-v2/README.md (hunk 1 above)
Comment (kdenhartog) [reply features/0587-encryption-envelope-v2/README.md:24]: Looks good to me based on what we have in the DIDCommV2 spec at the moment.

File: features/0587-encryption-envelope-v2/README.md (hunk 2 above)
Comment (troyronda) [reply features/0587-encryption-envelope-v2/README.md:56]: Resolved in https://github.com/decentralized-identity/didcomm-messaging/pull/168.

**Comment by andrewwhitehead** [comment]
It's a bit verbose, but I think another option would be to wrap the current agent message format in a JWM, to avoid having to do content type detection. For example:

```jsonc
{
    "id": "518be002-de8e-456e-b3d5-8fe472477a86",
    "type": "https://didcomm.org/compat/1.0/agent_message",
    "from": "did:example:alice",
    "to": ["did:example:bob"],
    "created_time": 1516269022,
    "expires_time": 1516385931,
    "body": {
        "@type": "https://didcomm.org/trust_ping/1.0/ping",
        "@id": "518be002-de8e-456e-b3d5-8fe472477a86",
        "~timing": {
            "out_time": "2018-12-15 04:29:23Z",
            "expires_time": "2018-12-15 05:29:23Z",
            "delay_milli": 0
        },
        "comment": "Hi. Are you listening?",
        "response_requested": true
    }
}
```

**Comment by troyronda** [comment]
@swcurran @TelegramSam @andrewwhitehead @llorllale I have updated this PR to state that the `cty` property of the JWE will distinguish DIDComm v1 vs DIDComm v2.

**Review by dhh1128 — APPROVED** [review]

**Comment by dhh1128** [comment]
LGTM.

**Review by kdenhartog — APPROVED** [review]
I like the new `CTY` answer for handling DIDCommV1 vs V2. It's such an obvious and simple solution. Great idea!

**Review by swcurran — APPROVED** [review]

**Pull request merged**